src/main/java/ch/njol/skript/expressions/ExprElement.java

```diff
@@ -29,87 +29,148 @@
+	private ElementType type;
 
 	@Override
 	@SuppressWarnings("unchecked")
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-		expr = LiteralUtils.defendExpression(exprs[2]);
-		type = ElementType.values()[parseResult.mark];
-		number = (Expression<Number>) (type == ElementType.ORDINAL ? exprs[0]: exprs[1]);
+		ElementType[] types = PATTERNS.getInfo(matchedPattern);
+		expr = LiteralUtils.defendExpression(exprs[exprs.length - 1]);
+		switch (types[0]) {
+			case RANGE:
+				endIndex = (Expression<Number>) exprs[1];
+				//$FALL-THROUGH$
+			case FIRST_X_ELEMENTS:
+			case ORDINAL:
+				startIndex = (Expression<Number>) exprs[0];
+				break;
+			default:
+				startIndex = null;
+				break;
+		}
+		type = types[parseResult.hasTag("last") ? 1 : 0];
 		return LiteralUtils.canInitSafely(expr);
 	}
 
 	@Override
 	@Nullable
 	protected Object[] get(Event event) {
-		Iterator<?> iter = expr.iterator(event);
-		if (iter == null || !iter.hasNext())
+		Iterator<?> iterator = expr.iterator(event);
+		if (iterator == null || !iterator.hasNext())
 			return null;
 		Object element = null;
+		int startIndex = 0;
+		int endIndex = 0;
+		if (this.startIndex != null) {
+			Number number = this.startIndex.getSingle(event);
+			if (number == null)
+				return null;
+			startIndex = number.intValue();
+		}
+		if (this.endIndex != null) {
+			Number number = this.endIndex.getSingle(event);
+			if (number == null)
+				return null;
+			endIndex = number.intValue();
+		}
 		switch (type) {
-			case FIRST:
-				element = iter.next();
+			case FIRST_ELEMENT:
+				element = iterator.next();
 				break;
-			case LAST:
-				element = Iterators.getLast(iter);
+			case LAST_ELEMENT:
+				element = Iterators.getLast(iterator);
+				break;
+			case RANDOM:
+				Object[] allElements = Iterators.toArray(iterator, Object.class);
+				element = CollectionUtils.getRandom(allElements);
 				break;
 			case ORDINAL:
-				assert this.number != null;
-				Number number = this.number.getSingle(event);
-				if (number == null)
+				if (startIndex <= 0)
 					return null;
-				try {
-					element = Iterators.get(iter, number.intValue() - 1);
-				} catch (IndexOutOfBoundsException exception) {
+				Iterators.advance(iterator, startIndex - 1);
+				if (!iterator.hasNext())
 					return null;
-				}
-				break;
-			case RANDOM:
-				Object[] allIterValues = Iterators.toArray(iter, Object.class);
-				element = CollectionUtils.getRandom(allIterValues);
+				element = iterator.next();
 				break;
 			case TAIL_END_ORDINAL:
-				allIterValues = Iterators.toArray(iter, Object.class);
-				assert this.number != null;
-				number = this.number.getSingle(event);
-				if (number == null)
-					return null;
-				int ordinal = number.intValue();
-				if (ordinal <= 0 || ordinal > allIterValues.length)
+				allElements = Iterators.toArray(iterator, Object.class);
+				if (startIndex <= 0 || startIndex > allElements.length)
```

Comment:
can we check startIndex before creating allElements? we don't want to waste time exhausting the iterator if we don't even use it